lib/server.dart

```diff
@@ -0,0 +1,112 @@
+import 'dart:convert';
+import 'package:meta/meta.dart';
+import 'package:xml/xml.dart';
+import 'package:xml_rpc/src/converter_extension.dart';
+
+import 'src/common.dart';
+import 'src/converter.dart';
+
+/// A [MethodFailureSignature] contains the name, parameters, and error from any exception or error during a function call
+///
+/// Name of the method being called
+/// Parameters that were passed in
+/// Exception or error that occurred
+typedef MethodFailureSignature = Fault Function(String, List<dynamic>, dynamic);
+
+/// A [XmlRpcHandler] handles the handling RPC functions along with marshalling the arguments and results to / from XMLRPC spec
+///
+/// Has a set of [codecs] for encoding and decoding the datatypes
+/// To register a method or function with the dispatcher call [registerFunction] along with the function and name
+class XmlRpcHandler {
+  /// The [codecs] used for encoding and decoding
+  List<Codec> codecs;
+
+  /// The function registry
+  final Map<String, Function> methods;
+
+  /// A function that gets called on a method failure with the following data
+  ///
+  /// Name of the method being called
+  /// Parameters that were passed in
+  /// Exception or error that occurred
+  MethodFailureSignature methodFailureHandler;
+
+  /// The error code for a method failure
+  final int methodFailureCode;
+
+  /// The error code for a missing method
+  final int noExistingMethodCode;
+
+  /// Creates a [XmlRpcHandler] with the set of [codecs] for encoding and decoding
+  XmlRpcHandler({
+    @required this.methods,
+    List<Codec> codecs,
+    MethodFailureSignature methodFailureHandler,
+    this.methodFailureCode,
+    this.noExistingMethodCode,
+  }) : codecs = codecs ?? standardCodecs {
+    this.methodFailureHandler =
+        methodFailureHandler ?? defaultMethodFailureHandler;
+  }
+
+  Fault defaultMethodFailureHandler(
+          String method, List<dynamic> params, dynamic error) =>
+      Fault(methodFailureCode,
+          'Dispatching $method, with params $params failed with error $error');
+
+  /// Marshalls the [data] from XML to Dart types, and then dispatches the function, and marshals the return value back into the XMLRPC format
+  Future<XmlDocument> handle(XmlDocument document) async {
+    final methodCall = document.findElements('methodCall').first;
+    final methodName = methodCall.findElements('methodName').first.text;
+    var returnValue;
+    if (methods.containsKey(methodName)) {
+      final params = methodCall.findElements('params');
+      final parsedArgs = params.isNotEmpty
+          ? params.first
+              .findElements('param')
+              .map((arg) => arg.findElements('value').first)
```

Comment:
Thanks for the reminder that `.first` can throw if list is empty. I'm sorry for the unnecessary change I requested.

Here a StateError could be thrown if no `<value>`...

Moreover decoders can throw ArgumentError ; we should try-catch those decoding/argument issues.

Reply:
Also, I'm adding handling of encoding errors